lib/Echidna/Solidity.hs

```diff
@@ -22,37 +23,46 @@ import EVM
   (VM, VMResult(..), contract, gas, loadContract, replaceCodeOfSelf, resetState, state)
 import EVM.Concrete (Blob(..))
 import EVM.Exec     (exec, vmForEthrunCreation)
-import EVM.Solidity (abiMap, contractName, creationCode, methodInputs, methodName, readSolc)
+import EVM.Solidity (abiMap, contractName, creationCode, methodInputs, methodName, readSolc, SolcContract)
 
 data EchidnaException = CompileFailure
                       | NoContracts
                       | TestArgsFound Text
+                      | ContractNotFound Text
 
 instance Show EchidnaException where
-  show CompileFailure    = "Couldn't compile given file"
-  show NoContracts       = "No contracts found in given file"
-  show (TestArgsFound t) = "Test " ++ show t ++ " has arguments, aborting"
+  show CompileFailure       = "Couldn't compile given file"
+  show NoContracts          = "No contracts found in given file"
+  show (ContractNotFound c) = "Given contract " ++ show c ++ " not found in given file"
+  show (TestArgsFound t)    = "Test " ++ show t ++ " has arguments, aborting"
 
 instance Exception EchidnaException
 
-loadSolidity :: (MonadIO m, MonadThrow m) => FilePath -> m (VM, [SolSignature], [Text])
-loadSolidity f = liftIO solc >>= \case
+selectContract :: (MonadThrow m) => [SolcContract] -> Maybe Text -> m SolcContract
```

Comment:
This isn't quite the abstraction boundary I'd draw. How would you feel about something of type `(MonadIO m, MonadThrow m) => FilePath -> Maybe Text -> m SolcContract`, and then your logic can be something like `let choose Nothing x = True; choose (Just s) x = x ^. contractName == s in liftIO solc >>= \case [...] Just m -> findOf (_1 . each) (choose n) m [...]`

Reply:
that would couple "reading all contracts from file" and "choosing the one to test" together. then they can't be used separately. makes the function more complex and the API less flexible.

how about having a function `loadContracts :: (MonadIO m, MonadThrow m) => FilePath -> m [SolcContract]`.  `(MonadIO m, MonadThrow m) => FilePath -> Maybe Text -> m SolcContract` would then simply be a composition of `loadContracts` and `selectContract`

Reply:
How do you see the average Echidna user using `selectContract`? I'm not sure what analysis would require such a function. I could see the utility of `loadContracts`, and don't have a strong preference between that and the function I proposed above (mildly prefer the latter I think?), but I'm unconvinced we should export a function for selecting a contract from a list thereof in a `MonadThrow`

Reply:
what do you think of `selectContract :: [SolcContract] -> Maybe Text -> Either EchidnaException SolcContract`? that would get rid of the `MonadThrow`. though the `MonadThrow` makes calling `selectContract` from `loadSolidity` a bit nicer.

i think of `selectContract` as a function encapsulating the strategy of selecting the contract to test. since it has two distinct failure cases we want to communicate we can't have it return `Maybe SolcContract` which would be preferable.

no need to export `selectContract` though i think extracting it into a function makes both functions easier to understand.

Reply:
> How do you see the average Echidna user using `selectContract`?

i'd like to enable the user to make that choice.
if we export `(MonadIO m, MonadThrow m) => FilePath -> Maybe Text -> m SolcContract` then we make a decision for the user that every time they want to read contracts from a file one contract is chosen using our strategy.
if the user wants something else they have to reimplement half of the function themselves.
if they want to use just the strategy but load contracts another way they have to reimplement again.

not saying that any choice should be enabled for the user. but if it's as straightforward as in this case i think it's worth doing.

big fan of APIs that export a set of small simple functions that each do one thing and can be combined in flexible unforeseen ways. reflects the reality of how people use APIs better imho.

Reply:
I think I'm just unclear on when the user would want something not covered by `(MonadIO m, MonadThrow m) => FilePath -> Maybe Text -> m SolcContract`. What kind of echidna testing does this not suffice for? I'd be happy to provide more granular functionality for users that require it, but I think I'd need to see some code that did first.